pulp_2to3_migration/app/plugin/rpm/comps_utils.py

```diff
@@ -0,0 +1,174 @@
+"""
+Convert data from mongo to libcomps
+
+"""
+
+import libcomps
+
+from pulp_rpm.app.comps import dict_to_strdict, list_to_idlist
+
+
+def langpacks_to_libcomps(obj):
+    """
+    Convert standard dict to langpacks libcomps StrDict type object.
+
+    Args:
+        obj(Pulp2PackageLangpacks): Pulp2PackageLangpacks
+
+    Returns:
+        strdict: a libcomps StrDict
+
+    """
+    strdict = libcomps.StrDict()
+    for dct in obj.matches:
+        strdict[dct['name']] = dct['install']
+    return strdict
+
+
+def pkg_cat_to_libcomps(obj):
+    """
+    Convert PackageCategory object to libcomps Category object.
+
+    Args:
+        obj(Pulp2PackageCategory): Pulp2PackageCategory
+
+    Returns:
+        group: libcomps.Category object
+
+    """
+    cat = libcomps.Category()
+
+    cat.id = obj.package_category_id
+    cat.name = obj.name
+    cat.desc = obj.description
+    cat.display_order = obj.display_order
+    cat.group_ids = list_to_idlist(_packages_to_grplist(obj.packagegroupids))
+    cat.desc_by_lang = dict_to_strdict(obj.desc_by_lang)
+    cat.name_by_lang = dict_to_strdict(obj.name_by_lang)
+
+    return cat
+
+
+def pkg_grp_to_libcomps(obj):
+    """
+    Convert PackageGroup object to libcomps Group object.
+
+    Args:
+        obj(Pulp2PackageGroup): Pulp2PackageGroup
+
+    Returns:
+        group: libcomps.Group object
+
+    """
+    group = libcomps.Group()
+
+    group.id = obj.package_group_id
+    group.default = obj.default
+    group.uservisible = obj.user_visible
+    group.display_order = obj.display_order
+    group.name = obj.name
+    group.desc = obj.description
+    group.packages = _list_to_pkglist(obj.packages)
+    group.biarchonly = obj.biarch_only
+    group.desc_by_lang = dict_to_strdict(obj.desc_by_lang)
+    group.name_by_lang = dict_to_strdict(obj.name_by_lang)
+
+    return group
+
+
+def pkg_env_to_libcomps(obj):
+    """
+    Convert PackageEnvironment object to libcomps Environment object.
+
+    Args:
+        obj(Pulp2PackageEnvironment): Pulp2PackageEnvironment
+
+    Returns:
+        group: libcomps.Environment object
+
+    """
+    env = libcomps.Environment()
+
+    env.id = obj.package_environment_id
+    env.name = obj.name
+    env.desc = obj.description
+    env.display_order = obj.display_order
+    env.group_ids = list_to_idlist(_packages_to_grplist(obj.group_ids))
+    result = _packages_to_optionlist(obj.option_ids)
+    env.option_ids = list_to_idlist(result)
+    env.desc_by_lang = dict_to_strdict(obj.desc_by_lang)
```

Comment:
This is not a question about this PR per se, but does ```StrDict``` have any special functionality that would prevent (theoretically) changing the bindings to just perform this conversion to and from Python dictionaries automatically?

If there's nothing special about them, and it's just a C implementation of a List or Dictionary, then it might be possible to (eventually) clean up the bindings with automatic conversions to and from Python-native types.

Reply:
@dralley, IIRC, StrDict is a type introduced by libcomps, it tries to behave like normal dict and string at the same time. Unfortunately, some dict operations were not supported. 

We definitely need just a dict and we don't use any special functionality if they provide. We never use it as a string, for example.
I think libcomps bindings  are auto-generated. I guess we can patch those but I'm not sure which one is the lesser of two evils: having the conversion in pulp or there.

Reply:
It's not autogenerated, they use CPython extensions, so it is less like "bindings" and more like "Python types implemented in C that don't don't really act like normal Python types".